src/aggregator/integration/one_client_passthru_test.go

```diff
@@ -0,0 +1,168 @@
+
+	// Clock setup.
+	var lock sync.RWMutex
+	now := time.Now().Truncate(time.Hour)
+	getNowFn := func() time.Time {
+		lock.RLock()
+		t := now
+		lock.RUnlock()
+		return t
+	}
+	setNowFn := func(t time.Time) {
+		lock.Lock()
+		now = t
+		lock.Unlock()
+	}
+	clockOpts := clock.NewOptions().SetNowFn(getNowFn)
+	serverOpts = serverOpts.SetClockOptions(clockOpts)
+
+	// Placement setup.
+	numShards := 1024
+	cfg := placementInstanceConfig{
+		instanceID:          serverOpts.InstanceID(),
+		shardSetID:          serverOpts.ShardSetID(),
+		shardStartInclusive: 0,
+		shardEndExclusive:   uint32(numShards),
+	}
+	instance := cfg.newPlacementInstance()
+	placement := newPlacement(numShards, []placement.Instance{instance})
+	placementKey := serverOpts.PlacementKVKey()
+	placementStore := serverOpts.KVStore()
+	require.NoError(t, setPlacement(placementKey, placementStore, placement))
+
+	// Create server.
+	testServer := newTestServerSetup(t, serverOpts)
+	defer testServer.close()
+
+	// Start the server.
+	log := testServer.aggregatorOpts.InstrumentOptions().Logger()
+	log.Info("test one client sending of passthrough metrics")
+	require.NoError(t, testServer.startServer())
+	log.Info("server is now up")
+	require.NoError(t, testServer.waitUntilLeader())
+	log.Info("server is now the leader")
+
+	var (
+		idPrefix = "full.passthru.id"
+		numIDs   = 10
+		start    = getNowFn()
+		stop     = start.Add(10 * time.Second)
+		interval = 2 * time.Second
+	)
+	client := testServer.newClient()
+	require.NoError(t, client.connect())
+	defer client.close()
+
+	ids := generateTestIDs(idPrefix, numIDs)
+	metadataFn := func(idx int) metadataUnion {
+		return metadataUnion{
+			mType:               passthroughMetadataType,
+			passthroughMetadata: policy.NewStoragePolicy(2*time.Second, xtime.Second, time.Hour),
+		}
+	}
+	dataset := mustGenerateTestDataset(t, datasetGenOpts{
+		start:        start,
+		stop:         stop,
+		interval:     interval,
+		ids:          ids,
+		category:     passthroughMetric,
+		typeFn:       constantMetricTypeFnFactory(metric.GaugeType),
+		valueGenOpts: defaultValueGenOpts,
+		metadataFn:   metadataFn,
+	})
+
+	for _, data := range dataset {
+		setNowFn(data.timestamp)
+		for _, mm := range data.metricWithMetadatas {
+			require.NoError(t, client.writePassthroughMetricWithMetadata(mm.metric.passthrough, mm.metadata.passthroughMetadata))
+		}
+		require.NoError(t, client.flush())
+
+		// Give server some time to process the incoming packets.
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	// Move time forward and wait for flushing to happen.
+	finalTime := stop.Add(time.Minute + 2*time.Second)
+	setNowFn(finalTime)
+	time.Sleep(2 * time.Second)
+
+	// Stop the server.
+	require.NoError(t, testServer.stopServer())
+	log.Info("server is now down")
+
+	// Validate results.
+	expected := computeExpectedPassthroughResults(t, dataset)
+	actual := testServer.sortedResults()
+	require.Equal(t, expected, actual)
```

Comment:
is this going to flake sometimes -- e.g. in case of retries or something?

i'd expect this to either dedupe actual, or check expected contained in actual. not sure if that's required tho, you tell me.

Reply:
Good point. Adding a dedup function for both expected and actual results.